src/index.ts

```diff
@@ -0,0 +1,241 @@
+      mix: cmix,
+      hash: keccak256(Buffer.concat([s, cmix]))
+    }
+  }
+
+  cacheHash() {
+    return keccak256(Buffer.concat(this.cache))
+  }
+
+  headerHash(header: BlockHeader) {
+    return rlphash(header.slice(0, -2))
+  }
+
+  /**
+   * Loads the seed and cache given a block number.
+   */
+  loadEpoc(number: number, cb: Function) {
+    const epoc = getEpoc(number)
+
+    if (this.epoc === epoc) {
+      return cb()
+    }
+
+    this.epoc = epoc
+
+    if (!this.cacheDB) {
+      throw new Error('cacheDB needed')
+    }
+
+    // gives the seed the first epoc found
+    const findLastSeed = (
+      epoc: number,
+      cb2: (seed: Buffer, epoc: number) => void
+    ) => {
+      if (epoc === 0) {
+        return cb2(zeros(32), 0)
+      }
+
+      this.cacheDB!.get(epoc, this.dbOpts, (err, data) => {
+        if (!err) {
+          cb2(data.seed, epoc)
+        } else {
+          findLastSeed(epoc - 1, cb2)
+        }
+      })
+    }
+
+    this.cacheDB.get(epoc, this.dbOpts, (err, data) => {
+      if (!data) {
+        this.cacheSize = getCacheSize(epoc)
+        this.fullSize = getFullSize(epoc)
+
+        findLastSeed(epoc, (seed, foundEpoc) => {
+          this.seed = getSeed(seed, foundEpoc, epoc)
+          const cache = this.mkcache(this.cacheSize!, this.seed!)
+          // store the generated cache
+          this.cacheDB!.put(
+            epoc,
+            {
+              cacheSize: this.cacheSize,
+              fullSize: this.fullSize,
+              seed: this.seed,
+              cache: cache
+            },
+            this.dbOpts,
+            cb as any
+          )
+        })
+      } else {
+        // Object.assign(this, data)
+        this.cache = data.cache.map((a: Buffer) => {
+          return Buffer.from(a)
+        })
+        this.cacheSize = data.cacheSize
+        this.fullSize = data.fullSize
+        this.seed = Buffer.from(data.seed)
+        cb()
+      }
+    })
+  }
+
+  _verifyPOW(header: BlockHeader, cb: (valid: boolean) => void) {
+    const headerHash = this.headerHash(header.raw)
+    const number = bufferToInt(header.number)
+
+    this.loadEpoc(number, () => {
+      const a = this.run(headerHash, Buffer.from(header.nonce, 'hex'))
+      const result = new BN(a.hash)
+      cb(
+        a.mix.toString('hex') === header.mixHash.toString('hex') &&
+          TWO_POW256.div(new BN(header.difficulty)).cmp(result) === 1
+      )
+    })
+  }
+
+  verifyPOW(block: Block, cb: (valid: boolean) => void) {
+    let valid = true
+
+    // don't validate genesis blocks
+    if (block.header.isGenesis()) {
+      cb(true)
+      return
+    }
+
+    this._verifyPOW(block.header, (valid2) => {
+      valid = valid2
+
+      if (!valid) {
+        return cb(valid)
+      }
+
+      async.eachSeries(
+        block.uncleHeaders,
+        (uheader: BlockHeader, cb2: (valid: boolean) => void) => {
+          this._verifyPOW(uheader, (valid3) => {
+            valid = valid3
+            cb2(valid)
```

Comment:
Are you confident on this simplification? 😄  Find this pretty hard to decipher (the old version).